aws-lc-rs/src/key_transport.rs

```diff
@@ -0,0 +1,408 @@
+    ptr::LcPtr,
+    ptr::Pointer,
+};
+use aws_lc::{
+    EVP_PKEY_CTX_kem_set_params, EVP_PKEY_CTX_new, EVP_PKEY_CTX_new_id, EVP_PKEY_decapsulate,
+    EVP_PKEY_encapsulate, EVP_PKEY_get_raw_private_key, EVP_PKEY_get_raw_public_key,
+    EVP_PKEY_kem_new_raw_public_key, EVP_PKEY_kem_new_raw_secret_key, EVP_PKEY_keygen,
+    EVP_PKEY_keygen_init, EVP_PKEY, EVP_PKEY_KEM, NID_KYBER512_R3,
+};
+use std::os::raw::c_int;
+use std::ptr::null_mut;
+use zeroize::Zeroize;
+
+// Key lengths defined as stated on the CRYSTALS website:
+// https://pq-crystals.org/kyber/
+//
+const KYBER512_R3_SECRET_KEY_LENGTH: usize = 1632;
+const KYBER512_R3_CIPHERTEXT_LENGTH: usize = 768;
+const KYBER512_R3_PUBLIC_KEY_LENGTH: usize = 800;
+const KYBER512_R3_SHARED_SECRET_LENGTH: usize = 32;
+
+#[allow(non_camel_case_types)]
+#[derive(Clone, Debug, PartialEq)]
+enum KemAlgorithmID {
+    Kyber512_R3,
+}
+
+/// A KEM algorithm
+#[derive(Debug, PartialEq)]
+pub struct KemAlgorithm {
+    id: KemAlgorithmID,
+    secret_key_size: usize,
+    public_key_size: usize,
+    ciphertext_size: usize,
+    shared_secret_size: usize,
+}
+
+/// NIST Round 3 iteration of the Kyber-512 algorithm
+pub static KYBER512_R3: KemAlgorithm = KemAlgorithm {
+    id: KemAlgorithmID::Kyber512_R3,
+    secret_key_size: KYBER512_R3_SECRET_KEY_LENGTH,
+    public_key_size: KYBER512_R3_PUBLIC_KEY_LENGTH,
+    ciphertext_size: KYBER512_R3_CIPHERTEXT_LENGTH,
+    shared_secret_size: KYBER512_R3_SHARED_SECRET_LENGTH,
+};
+
+impl KemAlgorithmID {
+    #[inline]
+    fn nid(&self) -> i32 {
+        match self {
+            KemAlgorithmID::Kyber512_R3 => NID_KYBER512_R3,
+        }
+    }
+}
+
+impl KemAlgorithm {
+    #[inline]
+    fn secret_key_size(&self) -> usize {
+        self.secret_key_size
+    }
+
+    #[inline]
+    fn public_key_size(&self) -> usize {
+        self.public_key_size
+    }
+
+    #[inline]
+    fn cipher_text_size(&self) -> usize {
+        self.ciphertext_size
+    }
+
+    #[inline]
+    fn shared_secret_size(&self) -> usize {
+        self.shared_secret_size
+    }
+}
+
+/// A serializable private key usable with KEMs. This can be randomly generated with KemPrivateKey::generate
+/// or constructed from raw bytes.
+#[derive(Debug)]
+pub struct KemPrivateKey {
+    algorithm: &'static KemAlgorithm,
+    pkey: LcPtr<*mut EVP_PKEY>,
+    priv_key: Box<[u8]>,
+}
+
+impl KemPrivateKey {
+    /// Generate a new KEM private key for the given algorithm.
+    ///
+    /// # Errors
+    /// `error::Unspecified` when operation fails due to internal error.
+    ///
+    pub fn generate(alg: &'static KemAlgorithm) -> Result<Self, Unspecified> {
+        unsafe {
+            let kyber_key = kem_key_generate(alg.id.nid())?;
+            let mut priv_key_bytes = vec![0u8; alg.secret_key_size()];
+            if 1 != EVP_PKEY_get_raw_private_key(
+                *kyber_key,
+                priv_key_bytes.as_mut_ptr(),
+                &mut alg.secret_key_size(),
```

Comment:
This value needs to be stored in a stack variable:
```
    let mut secret_key_size = alg.secret_key_size();
    let mut priv_key_bytes = vec![0u8; secret_key_size];
    if 1 != EVP_PKEY_get_raw_private_key(
        kyber_key.as_const_ptr(),
        priv_key_bytes.as_mut_ptr(),
        &mut secret_key_size,
    ) {
        return Err(Unspecified);
    }
```
Do we need to verify that `secret_key_size` was not changed prior to returning?

Reply:
> Do we need to verify that secret_key_size was not changed prior to returning?

The key size would never be changed since we are passing in a defined constant length for the algorithm we are using.